staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_test.go

```diff
@@ -0,0 +1,126 @@
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	runtimetesting "k8s.io/apimachinery/pkg/runtime/testing"
+	"k8s.io/apimachinery/pkg/test"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/component-base/version"
+)
+
+func TestEmulatedStorageVersion(t *testing.T) {
+	cases := []struct {
+		name             string
+		scheme           *runtime.Scheme
+		binaryVersion    schema.GroupVersion
+		effectiveVersion version.EffectiveVersion
+		want             schema.GroupVersion
+	}{
+		{
+			name:             "pick compatible",
+			scheme:           GetTestScheme(),
+			binaryVersion:    v1beta1,
+			effectiveVersion: version.NewEffectiveVersion("1.32"),
+			want:             v1alpha1,
+		},
+		{
+			name:             "alpha has been replaced, pick binary version",
+			scheme:           GetTestSchemeWithReplacedAlpha(),
+			binaryVersion:    v1beta1,
+			effectiveVersion: version.NewEffectiveVersion("1.32"),
+			want:             v1beta1,
+		},
+	}
+
+	for _, tc := range cases {
+		test.TestScheme()
+		t.Run(tc.name, func(t *testing.T) {
+			found, err := emulatedStorageVersion(tc.binaryVersion, &CronJob{}, tc.effectiveVersion, tc.scheme)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if found != tc.want {
+				t.Errorf("got %v; want %v", found, tc.want)
+			}
+		})
+	}
+}
+
+var internalGV = schema.GroupVersion{Group: "workload.example.com", Version: runtime.APIVersionInternal}
+var v1alpha1 = schema.GroupVersion{Group: "workload.example.com", Version: "v1alpha1"}
+var v1beta1 = schema.GroupVersion{Group: "workload.example.com", Version: "v1beta1"}
+
+type CronJobWithReplacement struct {
+	A string `json:"A,omitempty"`
+	B int    `json:"B,omitempty"`
+}
+
+func (*CronJobWithReplacement) GetObjectKind() schema.ObjectKind { panic("not implemented") }
+func (*CronJobWithReplacement) DeepCopyObject() runtime.Object {
+	panic("not implemented")
+}
+
+func (in *CronJobWithReplacement) APILifecycleIntroduced() (major, minor int) {
+	return 1, 30
+}
+
+func (in *CronJobWithReplacement) APILifecycleReplacement() schema.GroupVersionKind {
+	return schema.GroupVersionKind{Group: "rbac.authorization.k8s.io", Version: "v1", Kind: "ClusterRole"}
+}
+
+type CronJob struct {
+	A string `json:"A,omitempty"`
+	B int    `json:"B,omitempty"`
+}
+
+func (*CronJob) GetObjectKind() schema.ObjectKind { panic("not implemented") }
+func (*CronJob) DeepCopyObject() runtime.Object {
+	panic("not implemented")
+}
+
+func (in *CronJob) APILifecycleIntroduced() (major, minor int) {
+	return 1, 31
+}
+
+func GetTestScheme() *runtime.Scheme {
+	s := runtime.NewScheme()
+	s.AddKnownTypes(internalGV, &CronJob{})
+	s.AddKnownTypes(v1alpha1, &CronJob{})
+	s.AddKnownTypes(v1beta1, &CronJob{})
+	s.AddKnownTypeWithName(internalGV.WithKind("CronJob"), &CronJob{})
+	s.AddKnownTypeWithName(v1alpha1.WithKind("CronJob"), &CronJob{})
```

Comment:
Should we have the alpha cronjob being introduced at an earlier version rather than same version like in `CronJobWithReplacement`?